File: client/unity/GameMachine/Scripts/npc/WaypointRoute.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace GameMachine {
    public class WaypointRoute : MonoBehaviour {

        public enum Type {
            ReverseOnEnd,
            Circular,
            StopOnEnd
        }

        private List<Waypoint> waypoints = new List<Waypoint>();
        private int wpIndex = 0;
        private Waypoint currentWaypoint;

        public Type type;
        public bool showAtRuntime = false;
        

        void Start() {
            waypoints = transform.GetComponentsInChildren<Waypoint>().ToList();
            
            if (!showAtRuntime) {
                waypoints.ForEach(wp => Destroy(wp.GetComponent<MeshRenderer>()));
            }
            Reset();
        }

        public void Reset() {
            wpIndex = 0;
            currentWaypoint = waypoints.ToArray().ElementAt(wpIndex);
        }

        public int CurrentWaypointIndex() {
            return wpIndex;
        }

        public Waypoint CurrentWaypoint() {
            if (currentWaypoint == null) {
                return NextWaypoint();
            } else {
                return currentWaypoint;
            }
        }

        public Waypoint NextWaypoint() {
            if (wpIndex >= (waypoints.Count - 1)) {
                if (type == Type.ReverseOnEnd) {
                    waypoints.Reverse();
                    wpIndex = 0;
                } else if (type == Type.Circular) {
                    wpIndex = 0;
                }
            } else {
                wpIndex++;
            }

            return currentWaypoint = waypoints.ToArray().ElementAt(wpIndex);
        }
    }
}

File: server/java/server/src/main/java/io/gamemachine/unity/Path/WaypointRoute.java
package io.gamemachine.unity.Path;

import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by chris on 3/1/2016.
 */
public class WaypointRoute {

    public enum Type {
        ReverseOnEnd,
        Circular,
        Static
    }


    private List<Waypoint> waypoints = new ArrayList<Waypoint>();
    public int wpIndex = 0;
    private Waypoint currentWaypoint;

    public Type type;
    public boolean showAtRuntime = false;

    public WaypointRoute(List<Waypoint> waypoints) {
        Collections.sort(waypoints, (Waypoint o1, Waypoint o2) ->
                o1.order - o2.order);

        this.waypoints = waypoints;
    }

    public void reset() {
        wpIndex = 0;
        currentWaypoint = waypoints.get(wpIndex);
    }

    public int currentWaypointIndex() {
        return wpIndex;
    }

    public Waypoint CurrentWaypoint() {
        if (currentWaypoint == null) {
            return nextWaypoint();
        } else {
            return currentWaypoint;
        }
    }

    public Waypoint nextWaypoint() {
        if (type == Type.Static) {
            return currentWaypoint = waypoints.get(wpIndex);
        }

        if (wpIndex >= (waypoints.size() - 1)) {
            if (type == Type.ReverseOnEnd) {
                waypoints = Lists.reverse(waypoints);
                wpIndex = 0;
            } else if (type == Type.Circular) {
                wpIndex = 0;
            }
        } else {
            wpIndex++;
        }

        return currentWaypoint = waypoints.get(wpIndex);
    }
}

